include/commissioner/commissioner.hpp

```diff
@@ -127,35 +125,137 @@ struct Config
 };
 
 /**
- * @brief Enumeration of Joiner Type for steering.
+ * @brief The base class defines Handlers of commissioner events.
+ *
+ * Application should inherit this class and override the virtual
+ * functions to provide specific handler.
+ *
+ * @note keep in mind that those handlers will be called in another threads.
  *
  */
-enum class JoinerType
+class CommissionerHandler
 {
-    kMeshCoP = 0, ///< Conventional non-CCM joiner.
-    kAE,          ///< CCM AE joiner.
-    kNMKP         ///< CCM NMKP joiner.
-};
+public:
+    /**
+     * The function notifies the start of a joining request from given joiner.
+     *
+     * @param[in]  aJoinerId  A joiner ID.
+     *
+     * @return PSKd of the joiner. An empty PSKd indicates that the joiner is not
+     *         enabled.
+     *
+     */
+    virtual std::string OnJoinerRequest(const ByteArray &aJoinerId)
```

Comment:
```suggestion
    virtual std::string OnJoinerRequest(const ByteArray &aJoinerId) = 0;
```

What about use abstract functions so that users have to override these methods?

Reply:
There are too concerns for not using pure virtual functions:
1. The user is very likely to define this dummy handler if he/she is not going to use Thread 1.1 Commissioning (for  example, the user just want the Thread Network Management functionalities).
2. When binding to java, technically, we cannot enforce a Java class implementing every virtual functions at here. There will be a segment fault If we define them pure virtual here but a Java class doesn't implement them.

Reply:
For 1), it does not seems to be a serious issue to me. Nowadays, most IDEs can generate default implementations for unimplemented methods. This also have the advantage that the user can explicitly see what callbacks do `CommissionerHandler` provide.

For 2), can we just make sure all pure virtual functions are implemented since we own the Java code?

Reply:
> can we just make sure all pure virtual functions are implemented since we own the Java code?

We don't provide a `CommissionerHandler` implementation in Java. But I turns to agree we should make them pure virtual.